### PR TITLE
Optional TreeItem Checkbox

### LINKED ISSUE #116141: Allow TreeItems to have optional checkboxes
I'm all for trying to avoid webviews as much as possible, especially when tree views are hard to make work in a webview. This proposal allows for treeviews to have checkboxes:

image
(but better styling of course 😄)

#### Scenarios

* To-do list extensions
* Applying filters where an 'Edit' button drops you into the experience above where checkboxes show up
* Crafting `.gitignore`'s or other 'ignores' where you go down the filesystem and uncheck the files you want to ignore

#### Proposal

* TreeItem has new optional property called `checkboxState` similar to `collapsibleState` which takes in a `TreeItemCheckBoxState`:
```ts
        export class TreeItem {
            // ...

            /**
             * [TreeItemCheckboxState](#TreeItemCheckboxState) of the tree item.
             */
            checkboxState?: TreeItemCheckboxState;

            // ...
        }

       /**
	 * Checkbox state of the tree item
	 */
        export enum TreeItemCheckboxState{
		/**
		 * Determines an item can be neither checked nor unchecked. Implies it has no checkbox.
		 */
		None = 0,
		/**
		 * Determines an item is checked
		 */
		Checked = 1,
		/**
		 * Determines an item is unchecked
		 */
		Unchecked = 2
	}

```
* `TreeDataProvider` will have a new optional event `onDidChangeTreeCheckbox` which will fire when the checkbox of a `TreeItem` is checked or unchecked:
```ts
        /**
	 * A data provider that provides tree data
	 */
	export interface TreeDataProvider<T> {
                // ...

                /**
		 * An optional event to signal that an element or root has either been checked or unchecked.
		 */
		onDidChangeTreeCheckbox?: Event<ChangeTreeCheckboxEvent<T>>;

                // ...
        }

        export interface ChangeTreeCheckboxEvent {
                /**
		 * The item that was checked or unchecked.
		 */
		readonly item: <T>;

		/**
		 * Removed workspace folders.
		 */
		readonly newState: TreeItemCheckboxState;
        }
```

### PR DESCRIPTION
**Pull request by benibenj**

### CODE REVIEW

File: src/vscode-dts/vscode.proposed.treeItemCheckbox.d.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		* An optional event to signal that an element or root has either been checked or unchecked.
+		*/
+		onDidChangeTreeCheckbox?: Event<ChangeTreeCheckboxEvent>;
+	}
+
+	export interface ChangeTreeCheckboxEvent {
+		/**
+		* The item that was checked or unchecked.
+		*/
+		readonly items: TreeItem2[];
```
Comment (jrieken) [inline src/vscode-dts/vscode.proposed.treeItemCheckbox.d.ts:44]: Use `T` over `TreeItem`, use tuples (T, CheckedState)

**Review by alexr00 — CHANGES_REQUESTED** [review]
This is looking very good!

**Review by alexr00 — DISMISSED** [review]

**Review by alexr00 — DISMISSED** [review]

File: src/vs/workbench/browser/parts/views/treeView.ts [hunk 2]
```diff
@@ -1187,6 +1206,21 @@ class TreeRenderer extends Disposable implements ITreeRenderer<ITreeItem, FuzzyS
 		disposableStore.add(toDisposable(() => this.treeViewsService.removeRenderedTreeItemElement(node)));
 	}
 
+	private renderCheckbox(node: ITreeItem, templateData: ITreeExplorerTemplateData, disposableStore: DisposableStore) {
+		if (node.checkboxChecked !== undefined) {
+			if (!templateData.checkbox) {
+				const checkbox = new TreeItemCheckbox(templateData.checkboxContainer, this.checkboxStateHandler, this.themeService);
+				templateData.checkbox = checkbox;
+				disposableStore.add({ dispose: () => { checkbox.dispose(); templateData.checkbox = undefined; } });
```
Comment (alexr00) [inline src/vs/workbench/browser/parts/views/treeView.ts:1214]: There's a `toDisposable` utility that you can use: https://github.com/microsoft/vscode/blob/9ef2ee5ce1abc7812bea03375b47af44ec7d46af/extensions/git/src/util.ts#L28-L30

**Review by alexr00 — APPROVED** [review]
Feel free to merge despite my comment!

**Pull request merged**